### PR TITLE
[FIX] throw warning when meshes are unequal but have compatible number of vertices

### LINKED ISSUE #5303: use less stringent check on meshes
- We should partly revert https://github.com/nilearn/nilearn/pull/5296 as it leads to failure of the doc build

In many places, we have made it simple to work with 'related' meshes:
 
- for example plot a surface image that uses a 'pial' mesh but on background surface image with an 'inflated' mesh

For a similar reason, the following is crashing on main

```python
    img = load_sample_motor_activation_image()
    fsaverage_meshes = load_fsaverage()
    surface_stat_image = SurfaceImage.from_volume(
        mesh=fsaverage_meshes["pial"],
        volume_img=img,
    )

    fsaverage = load_fsaverage("fsaverage5")
    destrieux = fetch_atlas_surf_destrieux()
    destrieux_atlas = SurfaceImage(
        mesh=fsaverage["inflated"],
        data={
            "left": destrieux["map_left"],
            "right": destrieux["map_right"],
        },
    )

    mask = destrieux_atlas
    for part in mask.data.parts:
        mask.data.parts[part] = mask.data.parts[part] == 34

    masker = SurfaceMasker(mask)
    masker.fit_transform(surface_stat_image)
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #5303

Changes proposed in this pull request:

- check if meshes are equal in most cases and fall back to checking if they have the same number of vertices if that fails

### CODE REVIEW

**Review by hndgzkn — COMMENTED** [review]
I think we can define mesh comparison as a class method rather then putting it in utils. My suggestion is something similar to some of the previous reviews.

We have some objects like `SurfaceImage`, `SurfaceMesh`, `PoliMesh`, etc. We define those classes/objects, but we do not take advantage of the facilities that object oriented design provides us.

There are cases we either make some comparisons or an operation that require only the data that is contained already in that object. (`mean`, `get_data`, `equals`, etc.) I think for such cases it would be good to consider defining/implementing those operations in the class itself rather than in utils.

Otherwise we create many utility functions difficult to maintain and also this requires multiple imports where we actually have the object itself that can do the operation.

Otherwise, I think it is Ok.

File: nilearn/surface/utils.py
Comment (man-shu) [inline nilearn/surface/utils.py:1]: Just one codecov complaint here.

**Review by man-shu — APPROVED** [review]
Otherwise LGTM!

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Comment by Remi-Gau** [comment]
> I think we can define mesh comparison as a class method rather then putting it in utils. My suggestion is something similar to some of the previous reviews.
> 
> We have some objects like `SurfaceImage`, `SurfaceMesh`, `PoliMesh`, etc. We define those classes/objects, but we do not take advantage of the facilities that object oriented design provides us.
> 
> There are cases we either make some comparisons or an operation that require only the data that is contained already in that object. (`mean`, `get_data`, `equals`, etc.) I think for such cases it would be good to consider defining/implementing those operations in the class itself rather than in utils.
> 
> Otherwise we create many utility functions difficult to maintain and also this requires multiple imports where we actually have the object itself that can do the operation.
> 
> Otherwise, I think it is Ok.

- will tag https://github.com/nilearn/nilearn/issues/5300 for this comment, but I'd prefer to tackle this one in another PR
